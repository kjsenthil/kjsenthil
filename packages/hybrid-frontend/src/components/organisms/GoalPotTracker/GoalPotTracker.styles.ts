import styled from 'styled-components';
import { Theme } from '../../atoms';

// eslint-disable-next-line import/prefer-default-export
export const GoalPotTrackerContainer = styled.div`
  ${({ theme }: { theme: Theme }) => `
    display: flex;
    flex-direction: column;
    gap: ${theme.spacing(2.5)}px;
    width: 100%;
  `}
`;
