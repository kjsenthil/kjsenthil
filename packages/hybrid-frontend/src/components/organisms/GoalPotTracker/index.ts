import GoalPotTracker from './GoalPotTracker';

export default GoalPotTracker;
export type { GoalPotTrackerProps } from './GoalPotTracker';
