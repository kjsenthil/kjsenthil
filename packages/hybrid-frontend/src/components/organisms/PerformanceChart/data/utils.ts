export interface GetPerformanceDataResponse {
  // All numbers have 2 decimal places
  Data: {
    Type: 'performance';
    Id: string;
    Attributes: {
      AccountValue: number;
      Performance: {
        Value: number;
        Percentage: number; // This is a float. 50% will be 50.00
      };
      Values: Array<{ Date: string; Value: number }>;
    };
    Links: {
      self: string;
    };
    Relationships: {
      contributions: {
        Links: {
          related: string;
        };
        Data: {
          Type: 'contributions';
          Id: string;
        };
      };
    };
  };
  Included: Array<{
    // All numbers have 3 decimal places
    Type: 'contributions';
    Id: string;
    Attributes: {
      TotalContributions: number;
      Contributions: Array<{ Date: string; NetContributionsToDate: number }>;
    };
    Links: {
      self: string;
    };
    Relationships: null;
  }>;
}

export interface TimeSeriesDatum {
  date: Date;
  value: number;
}

export interface PerformanceDatum extends TimeSeriesDatum {
  metadata?: Record<string, unknown>;
}

export interface ContributionDatum extends TimeSeriesDatum {
  metadata?: Record<string, unknown>;
}

export enum PerformanceChartPeriod {
  '1M' = '1m',
  '3M' = '3m',
  '6M' = '6m',
  '1Y' = '1y',
  'ALL_TIME' = 'alltime',
}

export function sortTimeSeries<TS extends TimeSeriesDatum>(
  timeSeriesData: TS[],
  direction: 'asc' | 'desc' = 'asc'
): TS[] {
  return [...timeSeriesData].sort((d1, d2) =>
    direction === 'asc'
      ? d2.date.getTime() - d1.date.getTime()
      : d1.date.getTime() - d2.date.getTime()
  );
}

export function getTimeSeriesMinMax<TS extends TimeSeriesDatum>(timeSeriesData: TS[]) {
  if (timeSeriesData.length === 0) {
    return {
      minDate: new Date(0),
      maxDate: new Date(0),
      minValue: 0,
      maxValue: 0,
    };
  }

  let minDate = timeSeriesData[0].date;
  let maxDate = timeSeriesData[0].date;
  let minValue = timeSeriesData[0].value;
  let maxValue = timeSeriesData[0].value;

  timeSeriesData.forEach(({ date, value }) => {
    if (date > maxDate) maxDate = date;
    if (date < minDate) minDate = date;
    if (value > maxValue) maxValue = value;
    if (value < minValue) minValue = value;
  });

  return {
    minDate,
    maxDate,
    minValue,
    maxValue,
  };
}

export function timeSeriesDateAccessor(d: TimeSeriesDatum) {
  return d.date;
}

export function timeSeriesValueAccessor(d: TimeSeriesDatum) {
  return d.value;
}
