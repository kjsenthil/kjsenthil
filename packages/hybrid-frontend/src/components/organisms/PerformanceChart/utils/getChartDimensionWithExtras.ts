import { ChartDimension, ChartDimensionWithExtras } from '../../../../config/chart';

/**
 * This function calculates and injects certain extra figures (like inner width
 * and inner height) to a chart dimension object. It returns a new chart
 * dimension object containing these extras.
 */
export default function getChartDimensionWithExtras(
  chartDimension: ChartDimension
): ChartDimensionWithExtras {
  return {
    ...chartDimension,
    innerWidth: chartDimension.width - chartDimension.margin.left - chartDimension.margin.right,
    innerHeight: chartDimension.height - chartDimension.margin.top - chartDimension.margin.bottom,
  };
}
