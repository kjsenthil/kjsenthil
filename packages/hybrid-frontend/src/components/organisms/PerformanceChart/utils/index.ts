/* eslint-disable import/prefer-default-export */
export { default as getChartDimensionWithExtras } from './getChartDimensionWithExtras';
