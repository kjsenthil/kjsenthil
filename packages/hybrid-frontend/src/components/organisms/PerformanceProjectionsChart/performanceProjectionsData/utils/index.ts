export { default as valueDefinedFactory } from './contributionsDefined';
export { default as mapDate } from './mapDate';

export * from './accessors';
export * from './minMax';
