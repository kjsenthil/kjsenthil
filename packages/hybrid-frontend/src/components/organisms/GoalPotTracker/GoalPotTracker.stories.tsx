import * as React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import { formatCurrency } from '../../../utils/formatters';
import GoalPotTracker, { GoalPotTrackerProps } from './GoalPotTracker';

export default {
  title: 'Digital Hybrid/Organisms/Goal Pot Tracker',
  component: GoalPotTracker,
} as Meta;

const Template: Story<GoalPotTrackerProps> = (args) => <GoalPotTracker {...args} />;

const currencyFormatter = (val: number) =>
  formatCurrency(val, {
    opts: {
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    },
  });

const lumpSumValue = 375000;
const projectionValue = 1500000;
const remainingValue = 100000;
const totalValue = lumpSumValue + projectionValue + remainingValue;

const defaultArgs: GoalPotTrackerProps = {
  title: 'Your retirement pot',
  potTotal: totalValue,
  progressBarProps: {
    currencyFormatter,
    progressBarData: [
      {
        legendProps: { title: 'Lump sum', value: lumpSumValue },
        progress: lumpSumValue / totalValue,
      },
      {
        legendProps: { title: 'From age 67 - 89', value: projectionValue },
        progress: projectionValue / totalValue,
      },
      {
        legendProps: { title: 'Remaining', value: remainingValue },
        progress: remainingValue / totalValue,
      },
    ],
  },
};

export const Default = Template.bind({});
Default.args = defaultArgs;
