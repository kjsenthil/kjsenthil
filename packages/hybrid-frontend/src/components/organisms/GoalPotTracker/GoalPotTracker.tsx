import * as React from 'react';
import { Box, Typography, useTheme } from '../../atoms';
import { ProgressBarWithLegend, ProgressBarWithLegendProps } from '../../molecules';
import { formatCurrency } from '../../../utils/formatters';
import { GoalPotTrackerContainer } from './GoalPotTracker.styles';

export interface GoalPotTrackerProps {
  title: string;
  potTotal: number | undefined;

  progressBarProps: Omit<ProgressBarWithLegendProps, 'progressBarBackgrounds'>;
}

export default function GoalPotTracker({ title, potTotal, progressBarProps }: GoalPotTrackerProps) {
  const theme = useTheme();

  // When the pot total is undefined, there will be no text after the title
  const formattedRetirementTotal =
    potTotal !== undefined
      ? formatCurrency(potTotal, {
          opts: {
            minimumFractionDigits: 0,
            maximumFractionDigits: 0,
          },
        })
      : '';

  let progressBarBackgrounds: string[] | undefined;
  if (progressBarProps.progressBarData.length === 1) {
    progressBarBackgrounds = [];
    if (progressBarProps.progressBarData[0].legendProps.value === undefined) {
      // This happens when there is no income input yet
      progressBarBackgrounds.push(`${theme.palette.tertiary.light2}`);
    } else {
      // This happens when there is only income input
      progressBarBackgrounds.push(theme.palette.tertiary.main);
    }
  }

  return (
    <GoalPotTrackerContainer>
      <Box>
        <Typography component="span" variant="sh1" color="primary" colorShade="dark2">
          {title}
          {formattedRetirementTotal ? ': ' : ''}
        </Typography>
        <Typography component="span" variant="sh1" color="tertiary">{`${
          formattedRetirementTotal ? `${formattedRetirementTotal}` : ''
        }`}</Typography>
      </Box>

      <Box>
        <ProgressBarWithLegend
          progressBarBackgrounds={progressBarBackgrounds}
          {...progressBarProps}
        />
      </Box>
    </GoalPotTrackerContainer>
  );
}
