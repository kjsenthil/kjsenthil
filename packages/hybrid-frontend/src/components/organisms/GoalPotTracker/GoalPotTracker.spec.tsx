import * as React from 'react';
import { renderWithTheme } from '@tsw/test-util';
import { formatCurrency } from '../../../utils/formatters';
import GoalPotTracker from './GoalPotTracker';

const progressBarData = [
  {
    legendProps: { title: 'Title A', value: 10 },
    progress: 0.1,
  },
  {
    legendProps: { title: 'Title B', value: 30 },
    progress: 0.3,
  },
  {
    legendProps: { title: 'Title C', value: 70 },
    progress: 0.7,
  },
];

describe('GoalPotTracker', () => {
  it('matches snapshot', () => {
    expect(
      renderWithTheme(
        <GoalPotTracker
          title="Main title"
          potTotal={100}
          progressBarProps={{ currencyFormatter: formatCurrency, progressBarData }}
        />
      )
    ).toMatchSnapshot();
  });
});
