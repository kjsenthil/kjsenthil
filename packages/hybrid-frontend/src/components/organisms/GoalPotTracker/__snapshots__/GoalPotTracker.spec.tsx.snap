// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalPotTracker matches snapshot 1`] = `
Object {
  "result": Object {
    "asFragment": [Function],
    "baseElement": <body>
      <div>
        <div
          class="sc-lcequs jWiuCR"
        >
          <div
            class="MuiBox-root MuiBox-root-1"
          >
            <span
              class="MuiTypography-root sc-jrQzUz jCZfJK MuiTypography-body1"
            >
              Main title
              : 
            </span>
            <span
              class="MuiTypography-root sc-jrQzUz UTINV MuiTypography-body1"
            >
              £100
            </span>
          </div>
          <div
            class="MuiBox-root MuiBox-root-2"
          >
            <div
              aria-valuenow="110.00000000000001"
              class="sc-iCfLBT fLrhme"
              role="progressbar"
            >
              <div
                class="sc-pVTma kWowoK"
              />
              <div
                class="sc-furvIG erDLzm"
              />
              <div
                class="sc-furvIG iDqOBC"
              />
              <div
                class="sc-furvIG avoAJ"
              />
            </div>
            <div
              class="MuiBox-root MuiBox-root-3"
              data-testid="Spacer"
            />
            <div
              class="sc-ikJzcn gZHwkl"
            >
              <div
                class="sc-jJoQpE isypUD"
                width="10"
              >
                <div
                  class="sc-bBHwJV kXmoUd"
                >
                  <div>
                    <h6
                      class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
                    >
                      TITLE A
                    </h6>
                    <div
                      class="sc-iwjezw hwFGdy"
                    >
                      <h6
                        class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                      >
                        £10.00
                      </h6>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="sc-jJoQpE bmnsXL"
                width="30"
              >
                <div
                  class="sc-bBHwJV kXmoUd"
                >
                  <div>
                    <h6
                      class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
                    >
                      TITLE B
                    </h6>
                    <div
                      class="sc-iwjezw hwFGdy"
                    >
                      <h6
                        class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                      >
                        £30.00
                      </h6>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="sc-jJoQpE hYOjDQ"
                width="70"
              >
                <div
                  class="sc-bBHwJV kXmoUd"
                >
                  <div>
                    <h6
                      class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
                    >
                      TITLE C
                    </h6>
                    <div
                      class="sc-iwjezw hwFGdy"
                    >
                      <h6
                        class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                      >
                        £70.00
                      </h6>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </body>,
    "container": <div>
      <div
        class="sc-lcequs jWiuCR"
      >
        <div
          class="MuiBox-root MuiBox-root-1"
        >
          <span
            class="MuiTypography-root sc-jrQzUz jCZfJK MuiTypography-body1"
          >
            Main title
            : 
          </span>
          <span
            class="MuiTypography-root sc-jrQzUz UTINV MuiTypography-body1"
          >
            £100
          </span>
        </div>
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <div
            aria-valuenow="110.00000000000001"
            class="sc-iCfLBT fLrhme"
            role="progressbar"
          >
            <div
              class="sc-pVTma kWowoK"
            />
            <div
              class="sc-furvIG erDLzm"
            />
            <div
              class="sc-furvIG iDqOBC"
            />
            <div
              class="sc-furvIG avoAJ"
            />
          </div>
          <div
            class="MuiBox-root MuiBox-root-3"
            data-testid="Spacer"
          />
          <div
            class="sc-ikJzcn gZHwkl"
          >
            <div
              class="sc-jJoQpE isypUD"
              width="10"
            >
              <div
                class="sc-bBHwJV kXmoUd"
              >
                <div>
                  <h6
                    class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
                  >
                    TITLE A
                  </h6>
                  <div
                    class="sc-iwjezw hwFGdy"
                  >
                    <h6
                      class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                    >
                      £10.00
                    </h6>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-jJoQpE bmnsXL"
              width="30"
            >
              <div
                class="sc-bBHwJV kXmoUd"
              >
                <div>
                  <h6
                    class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
                  >
                    TITLE B
                  </h6>
                  <div
                    class="sc-iwjezw hwFGdy"
                  >
                    <h6
                      class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                    >
                      £30.00
                    </h6>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-jJoQpE hYOjDQ"
              width="70"
            >
              <div
                class="sc-bBHwJV kXmoUd"
              >
                <div>
                  <h6
                    class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
                  >
                    TITLE C
                  </h6>
                  <div
                    class="sc-iwjezw hwFGdy"
                  >
                    <h6
                      class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                    >
                      £70.00
                    </h6>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>,
    "debug": [Function],
    "findAllByAltText": [Function],
    "findAllByDisplayValue": [Function],
    "findAllByLabelText": [Function],
    "findAllByPlaceholderText": [Function],
    "findAllByRole": [Function],
    "findAllByTestId": [Function],
    "findAllByText": [Function],
    "findAllByTitle": [Function],
    "findByAltText": [Function],
    "findByDisplayValue": [Function],
    "findByLabelText": [Function],
    "findByPlaceholderText": [Function],
    "findByRole": [Function],
    "findByTestId": [Function],
    "findByText": [Function],
    "findByTitle": [Function],
    "getAllByAltText": [Function],
    "getAllByDisplayValue": [Function],
    "getAllByLabelText": [Function],
    "getAllByPlaceholderText": [Function],
    "getAllByRole": [Function],
    "getAllByTestId": [Function],
    "getAllByText": [Function],
    "getAllByTitle": [Function],
    "getByAltText": [Function],
    "getByDisplayValue": [Function],
    "getByLabelText": [Function],
    "getByPlaceholderText": [Function],
    "getByRole": [Function],
    "getByTestId": [Function],
    "getByText": [Function],
    "getByTitle": [Function],
    "queryAllByAltText": [Function],
    "queryAllByDisplayValue": [Function],
    "queryAllByLabelText": [Function],
    "queryAllByPlaceholderText": [Function],
    "queryAllByRole": [Function],
    "queryAllByTestId": [Function],
    "queryAllByText": [Function],
    "queryAllByTitle": [Function],
    "queryByAltText": [Function],
    "queryByDisplayValue": [Function],
    "queryByLabelText": [Function],
    "queryByPlaceholderText": [Function],
    "queryByRole": [Function],
    "queryByTestId": [Function],
    "queryByText": [Function],
    "queryByTitle": [Function],
    "rerender": [Function],
    "unmount": [Function],
  },
}
`;
